cpp/cograph_recognition/cograph/Recognition.cpp

```diff
@@ -0,0 +1,549 @@
+                u ->addchild(z);
+                z ->addchild(x);
+                z->addchild(y);
+            }
+        }
+    }
+
+
+    int CographRecognition::Cograph_Recognition(NetworKit::Graph &graph){
+        CoNode *R = new CoNode(Type::ZEROONE, 1);
+        CoTree Tp(R);
+        T = &Tp;
+        T -> add(R);
+        G = graph;
+        vector<NetworKit::node>vertex;
+        vector<CoNode*>covertex;
+        map<NetworKit::node, int> pos;
+        int cnt = 0;
+        for(auto i : G.nodeRange()){
+            vertex.push_back(i);
+            pos[i] = cnt++;
+            CoNode *C = new CoNode(Type::VERTEX,i);
+            T -> add(C);
+            covertex.push_back(C);
+        }
+        for(auto i : G.nodeRange()){
+            vector<CoNode*> vec;
+            for(auto u : G.neighborRange(i)){
+               // cout<<i<<" "<<u<<endl;
+                vec.push_back(covertex[pos[u]]);
+            }
+            covertex[pos[i]] -> setoutEdges(vec);
+        }
+
+
+
+
+        if(cnt == 0){
+            T ->clear();
+            return 0;
+        }
+        if(cnt == 1){
+            R ->addchild(covertex[0]);
+            T -> clear();
+            return 0;
+        }
+        if(G.hasEdge(vertex[0], vertex[1])){
+            R ->addchild(covertex[0]);
+            R ->addchild(covertex[1]);
+        } else{
+            CoNode *N = new CoNode(Type::ZEROONE, 0);
+            T -> add(N);
+            R ->addchild(N);
+            N -> addchild(covertex[0]);
+            N -> addchild(covertex[1]);
+        }
+
+        covertex[0] -> add_to_graph();
+        covertex[1] -> add_to_graph();
+        CoNode* root = R;
+       // rec(root);
+
+        for(int i = 2; i < cnt; i++){
+            root = T -> getRoot();
+            Reset_All_CoNodes(root);
+            Mark(covertex[i]);
+            if(root -> Marked_or_not() == Marked::MARKED_AND_UNMARKED){//all nodes of T were marked and unmarked <=> R is marked and unmarked
+             //   cout<<"1111"<<" "<<i<<endl;
+                root -> addchild(covertex[i]);
+            } else if(mark_ever_count == 0){//
+              //  cout<<"2222"<<" "<<i<<endl;
+                if(root -> get_d() == 1){
+                    (root -> get_head_of_list_of_children()) -> addchild(covertex[i]);
+                } else{
+                    CoNode *R1 = new CoNode(Type::ZEROONE, 1);
+                    CoNode *R2 = new CoNode(Type::ZEROONE, 0);
+                    T -> add(R1);
+                    T -> add(R2);
+                    R1 -> addchild(R2);
+                    R2 -> addchild(root);
+                    R2 -> addchild(covertex[i]);
+                    T ->setRoot(R1);
+                    root = R1;
+                }
+            } else{
+              //  cout<<"3333 "<<i<<endl;
+                int error = 0;
+                CoNode* u = Find_Lowest(error);
+                if(error){
+                    T -> clear();
+                    return error;
```

Comment:
This should be reworked. In general, it's a really bad design to use error parameter like that, especially since C++ for quite a time allow for returning multiple values.

Reply:
maybe error is a bad name for it, it is only a type of given graph. i think i don't need to return multiple values, i only require this type.